File: tools/e2e-cli/src/tests/dist-tags.ts
import assert from 'assert';

import { TestContext, TestDefinition } from '../types';

async function publishPkg(ctx: TestContext, tempFolder: string, extraArgs: string[] = []) {
  await ctx.adapter.exec(
    { cwd: tempFolder },
    'publish',
    ...extraArgs,
    ...ctx.adapter.registryArg(ctx.registryUrl)
  );
}

async function bumpVersion(ctx: TestContext, tempFolder: string) {
  await ctx.adapter.exec(
    { cwd: tempFolder },
    'version',
    'minor',
    ...ctx.adapter.registryArg(ctx.registryUrl)
  );
}

async function testDistTags(ctx: TestContext): Promise<void> {
  const id = ctx.runId;

  // Test 1: list dist-tags
  const pkgName1 = `@foo/dt1-${id}`;
  const { tempFolder: tf1 } = await ctx.adapter.prepareProject(
    pkgName1,
    '1.0.0',
    ctx.registryUrl,
    ctx.port,
    ctx.token
  );
  await publishPkg(ctx, tf1);
  await bumpVersion(ctx, tf1);
  await publishPkg(ctx, tf1, ['--tag', 'beta']);
  const resp1 = await ctx.adapter.exec(
    { cwd: tf1 },
    'dist-tag',
    'ls',
    '--json',
    ...ctx.adapter.registryArg(ctx.registryUrl)
  );
  assert.strictEqual(resp1.stdout, 'beta: 1.1.0latest: 1.0.0');

  // Test 2: remove tag
  const pkgName2 = `@verdaccio/dt2-${id}`;
  const { tempFolder: tf2 } = await ctx.adapter.prepareProject(
    pkgName2,
    '1.0.0',
    ctx.registryUrl,
    ctx.port,
    ctx.token
  );
  await publishPkg(ctx, tf2);
  await bumpVersion(ctx, tf2);
  await publishPkg(ctx, tf2, ['--tag', 'beta']);
  const resp2 = await ctx.adapter.exec(
    { cwd: tf2 },
    'dist-tag',
    'rm',
    `${pkgName2}@1.1.0`,
    'beta',
    ...ctx.adapter.registryArg(ctx.registryUrl)
  );
  assert.strictEqual(resp2.stdout, `-beta: ${pkgName2}@1.1.0`);

  // Test 3: add tag
  const pkgName3 = `@verdaccio/dt3-${id}`;
  const { tempFolder: tf3 } = await ctx.adapter.prepareProject(
    pkgName3,
    '1.0.0',
    ctx.registryUrl,
    ctx.port,
    ctx.token
  );
  await publishPkg(ctx, tf3);
  await bumpVersion(ctx, tf3);
  await publishPkg(ctx, tf3);
  const resp3 = await ctx.adapter.exec(
    { cwd: tf3 },
    'dist-tag',
    'add',
    `${pkgName3}@1.1.0`,
    'alfa',
    ...ctx.adapter.registryArg(ctx.registryUrl)
  );
  assert.strictEqual(resp3.stdout, `+alfa: ${pkgName3}@1.1.0`);
}

export const distTagsTest: TestDefinition = {
  name: 'dist-tags',
  requires: ['dist-tag'],
  run: testDistTags,
};

File: tools/e2e-cli/src/e2e_tests/dist_tags.rs
use super::prelude::*;

pub(crate) fn test_dist_tags(ctx: &mut TestContext<'_>) -> Result<()> {
    let pkg1 = format!("@foo/dt1-{}", ctx.run_id);
    let tf1 = prepare_project(ctx, &pkg1, "1.0.0", BTreeMap::new(), BTreeMap::new())?;
    publish_pkg(ctx, &tf1)?;
    bump_version(ctx, &tf1, "minor")?;
    publish_pkg_extra(ctx, &tf1, vec!["--tag".into(), "beta".into()])?;
    let mut args = vec!["dist-tag".into(), "ls".into(), "--json".into()];
    args.extend(registry_arg(&ctx.adapter, &ctx.registry_url));
    let resp = exec_adapter(ctx, Some(&tf1), args)?;
    assert_eq_str(
        &normalize_dist_tag_list(&resp.stdout),
        "beta: 1.1.0\nlatest: 1.0.0",
    )?;

    let pkg2 = format!("@verdaccio/dt2-{}", ctx.run_id);
    let tf2 = prepare_project(ctx, &pkg2, "1.0.0", BTreeMap::new(), BTreeMap::new())?;
    publish_pkg(ctx, &tf2)?;
    bump_version(ctx, &tf2, "minor")?;
    publish_pkg_extra(ctx, &tf2, vec!["--tag".into(), "beta".into()])?;
    let mut args = vec![
        "dist-tag".into(),
        "rm".into(),
        format!("{pkg2}@1.1.0"),
        "beta".into(),
    ];
    args.extend(registry_arg(&ctx.adapter, &ctx.registry_url));
    let resp = exec_adapter(ctx, Some(&tf2), args)?;
    assert_eq_str(resp.stdout.trim(), &format!("-beta: {pkg2}@1.1.0"))?;

    let pkg3 = format!("@verdaccio/dt3-{}", ctx.run_id);
    let tf3 = prepare_project(ctx, &pkg3, "1.0.0", BTreeMap::new(), BTreeMap::new())?;
    publish_pkg(ctx, &tf3)?;
    bump_version(ctx, &tf3, "minor")?;
    publish_pkg(ctx, &tf3)?;
    let mut args = vec![
        "dist-tag".into(),
        "add".into(),
        format!("{pkg3}@1.1.0"),
        "alfa".into(),
    ];
    args.extend(registry_arg(&ctx.adapter, &ctx.registry_url));
    let resp = exec_adapter(ctx, Some(&tf3), args)?;
    assert_eq_str(resp.stdout.trim(), &format!("+alfa: {pkg3}@1.1.0"))?;
    Ok(())
}

fn normalize_dist_tag_list(output: &str) -> String {
    output
        .lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .collect::<Vec<_>>()
        .join("\n")
}
